compiler/src/parsing/well_formedness.re

```diff
@@ -367,27 +377,83 @@ let no_zero_denominator_rational = (errs, super) => {
   {errs, iterator};
 };
 
-let known_attributes = ["disableGC"];
+type known_attribute = {
+  name: string,
+  arity: int,
+};
+
+let known_attributes = [
+  {name: "disableGC", arity: 0},
+  {name: "externalName", arity: 1},
+];
+
+let valid_attributes = (errs, super) => {
+  let iter_attributes = (({txt, loc}, args)) => {
+    switch (List.find_opt(({name}) => name == txt, known_attributes)) {
+    | Some({arity}) when List.length(args) != arity =>
+      errs := [InvalidAttributeArity(txt, arity, loc), ...errs^]
+    | None => errs := [UnknownAttribute(txt, loc), ...errs^]
+    | _ => ()
+    };
+  };
 
-let no_unknown_attributes = (errs, super) => {
   let iter_expr = (self, {pexp_attributes: attrs} as e) => {
-    List.iter(
-      ({txt, loc}) =>
-        if (!List.mem(txt, known_attributes)) {
-          errs := [UnknownAttribute(txt, loc), ...errs^];
-        },
-      attrs,
-    );
+    List.iter(iter_attributes, attrs);
     super.expr(self, e);
   };
   let iter_toplevel = (self, {ptop_attributes: attrs} as top) => {
-    List.iter(
-      ({txt, loc}) =>
-        if (!List.mem(txt, known_attributes)) {
-          errs := [UnknownAttribute(txt, loc), ...errs^];
-        },
-      attrs,
-    );
+    List.iter(iter_attributes, attrs);
+    super.toplevel(self, top);
+  };
+  let iterator = {...super, expr: iter_expr, toplevel: iter_toplevel};
+  {errs, iterator};
+};
+
+let disallowed_attributes = (errs, super) => {
+  let iter_expr = (self, {pexp_desc: desc, pexp_attributes: attrs} as e) => {
+    switch (List.find_opt((({txt}, _)) => txt == "externalName", attrs)) {
+    | Some(({txt, loc}, _)) =>
+      errs :=
+        [
+          AttributeDisallowed(
+            "`externalName` is only allowed on top-level let bindings and `foreign` statements.",
+            loc,
+          ),
+        ]
+    | None => ()
+    };
+    super.expr(self, e);
+  };
+  let iter_toplevel =
+      (self, {ptop_desc: desc, ptop_attributes: attrs} as top) => {
+    switch (List.find_opt((({txt}, _)) => txt == "externalName", attrs)) {
+    | Some(({txt, loc}, _)) =>
+      switch (desc) {
+      | PTopLet(
+          _,
+          _,
+          _,
+          [
+            {
+              pvb_pat: {
+                ppat_desc:
+                  PPatVar(_) | PPatConstraint({ppat_desc: PPatVar(_)}, _),
+              },
+            },
+          ],
+        )
+      | PTopForeign(_) => ()
+      | _ =>
+        errs :=
+          [
+            AttributeDisallowed(
+              "`externalName` is only allowed on `foreign` statements and let bindings of a single name.",
```

Comment:
What do you mean by "let bindings of a single name"?

Reply:
It means you aren't allowed to do `@externalName let a = 1, b = 2` (or `@externalName let (a, b) = ...`). I couldn't think of a better way to say that 😢

Reply:
How about "let bindings that aren't a destructuring pattern" or something along those lines?

Reply:
We'd still need to cover the multiple binds case. Could just turn it into two separate error messages.